### PR TITLE
Include filtered columns in EXPLAIN (TYPE IO)

### PR DESCRIPTION
**Pull request by JamesRTaylor**
See issue here: https://github.com/prestosql/presto/issues/1120. This PR adds the columns that are filtered on taking the same approach as PlanPrinter.java. We use this information to track which columns are most frequently filtered to drive which columns should be clustered on. Who would be the right person to review, @martint?

### CODE REVIEW

**Comment by dain** [comment]
@Praveen2112  can you take a look at this one?

**Comment by Praveen2112** [comment]
@dain Sure

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 1]
```diff
@@ -249,13 +262,155 @@ public void testIoExplain()
...
+                                                        new FormattedDomain(
+                                                                false,
+                                                                ImmutableSet.of(
+                                                                        new FormattedRange(
+                                                                                new FormattedMarker(Optional.of("1"), EXACTLY),
+                                                                                new FormattedMarker(Optional.of("1"), EXACTLY)),
+                                                                        new FormattedRange(
+                                                                                new FormattedMarker(Optional.of("2"), EXACTLY),
+                                                                                new FormattedMarker(Optional.of("2"), EXACTLY)))))),
+                                        Sets.newLinkedHashSet(
```
Comment (Praveen2112) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]:
```suggestion
                                        ImmutableSet.of(
```

File: presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java [hunk 2]
```diff
@@ -600,19 +616,90 @@ protected Void visitPlan(PlanNode node, IoPlanBuilder context)
...
 
         @Override
-        public Void visitTableScan(TableScanNode node, IoPlanBuilder context)
+        public Void visitFilter(FilterNode node, IoPlanBuilder context)
+        {
+            return visitScanFilterAndProjectInfo(node, Optional.of(node), Optional.empty(), context);
+        }
+
+        @Override
+        public Void visitProject(ProjectNode node, IoPlanBuilder context)
```
Comment (Praveen2112) [inline presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java]: Can we directly visit the filterNode instead of going via project node ?

File: presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java [hunk 3]
```diff
@@ -600,19 +616,90 @@ protected Void visitPlan(PlanNode node, IoPlanBuilder context)
...
+            if (scanNode.isPresent() && filterNode.isPresent()) {
+                addConstraintsFromFilter(filterNode.get(), scanNode.get(), context);
+                return null;
+            }
+
+            sourceNode.accept(this, context);
+            return null;
+        }
+
+        private void addConstraintsFromFilter(FilterNode filterNode, TableScanNode tableScanNode, IoPlanBuilder context)
```
Comment (Praveen2112) [inline presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java]: Instead of passing the `FilterNode` here can we pass the  filter `Expression` ? So when we visit a normal TableScan we compute the column details like 
```
addConstraintsFromFilter(TRUE_LITERAL, tableScanNode, context)
```

When visiting a FilterNode we can check for its source and compute i/o based on its source

**Comment by JamesRTaylor** [comment]
@Praveen2112 - thanks for the review. Good suggestions. This new version is much simpler than the previous one.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java [hunk 4]
```diff
@@ -249,13 +262,289 @@ public void testIoExplain()
...
```
Comment (Praveen2112) [inline presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]:
```suggestion
                                        ImmutableSet.of
```

File: presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java [hunk 5]
```diff
@@ -600,19 +615,46 @@ protected Void visitPlan(PlanNode node, IoPlanBuilder context)
...
+            PlanNode source = node.getSource();
+            if (source instanceof TableScanNode) {
+                addInputTableConstraints(node.getPredicate(), (TableScanNode) source, context);
+                return null;
+            }
+
+            return processChildren(node, context);
+        }
+
+        private void addInputTableConstraints(Expression filterExpression, TableScanNode tableScanNode, IoPlanBuilder context)
```
Comment (Praveen2112) [inline presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java]: Can we move this method below the overridden methods ?

**Comment by JamesRTaylor** [comment]
@Praveen2112 - done. Thanks again for reviews.

File: presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java (hunk 4 above)
Comment (Praveen2112) [reply presto-hive/src/test/java/io/prestosql/plugin/hive/TestHiveIntegrationSmokeTest.java]: I guess we can change this to `ImmutableSet.of`

**Comment by martint** [comment]
What's the purpose of separating them into two sets (i.e., "constraints" and "filters"). For the purpose of describing how columns are being considered for filtering at the bottom level, I think just having them all in one set would be simpler and much easier to explain to users.

**Comment by JamesRTaylor** [comment]
The reasoning to separate partition constraints from other constraints are:

1. Easier post processing of io explain plan. The processes that analyze the explain plan don't easily know which columns are partition columns and they need to treat them differently. For example, we want to know which columns are most commonly filtered on so need to normalize to io explain plan results. We'd want to aggregate results together for two queries that only differ on which partitions are filtered (i.e. they have different date ranges in the query).
2. Keep format of JSON backward compatible. Existing processes may depend on only having partition constraints under the constraints element.

(1) could also be solved by adding an attribute or element to the JSON element to indicate that it's a partition constraint.

**Comment by JamesRTaylor** [comment]
Thinking about this more, I think the post processing of the io explain plan will be fine if it's combined into one set. There's still the issue around backward compatibility, but perhaps that's not an issue/goal?

**Comment by martint** [comment]
The old field used to be called `columns`. Now it's `columnConstraints` vs `columnFilters`, so that's already a backward incompatible change. I think we should just put them all under `columns` so format-wise it's still compatible.

Note that there's nothing that says those columns would be partition keys -- remember that the engine has no notion of such a concept. The list is just a description of the domain of the columns being considered for the query. In that sense, putting them all in the same set is in the spirit of that concept.

If tools do any interpretation of the results and the list of columns, they are likely aware of whether a column is a partition key or have access to metadata about that. They should be responsible to filter out those that are not interesting to them.

**Comment by JamesRTaylor** [comment]
Thanks for the review, @martint.  I've updated the PR based on your feedback.

File: presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java [hunk 6]
```diff
@@ -244,36 +247,36 @@ public boolean equals(Object obj)
...
-                        .add("columnConstraints", columnConstraints)
+                        .add("columns", columns)
                         .add("estimate", estimate)
                         .toString();
             }
         }
     }
 
-    public static class ColumnConstraint
+    public static class Column
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java]: I don't see a good reason for changing the name of this class, so I'd keep it as it was before. At a minimum, if we're changing it to make usage or meaning clearer, it should be done in a separate commit.

File: presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java [hunk 7]
```diff
@@ -650,6 +663,22 @@ else if (writerTarget instanceof CreateReference || writerTarget instanceof Inse
             return processChildren(node, context);
         }
 
+        private void addInputTableConstraints(TupleDomain<ColumnHandle> filterDomain, TableScanNode tableScanNode, IoPlanBuilder context)
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java]: Rename `tableScanNode` to just `tableScan`

File: presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java [hunk 8]
```diff
@@ -650,6 +663,22 @@ else if (writerTarget instanceof CreateReference || writerTarget instanceof Inse
...
+            TableMetadata tableMetadata = metadata.getTableMetadata(session, table);
+            TupleDomain<ColumnHandle> predicate = metadata.getTableProperties(session, table).getPredicate();
+            EstimatedStatsAndCost estimatedStatsAndCost = getEstimatedStatsAndCost(tableScanNode);
+            context.addInputTableColumnInfo(
+                    new IoPlan.TableColumnInfo(
+                            new CatalogSchemaTableName(
+                                    tableMetadata.getCatalogName().getCatalogName(),
+                                    tableMetadata.getTable().getSchemaName(),
+                                    tableMetadata.getTable().getTableName()),
+                            Sets.union(parseConstraints(table, predicate), parseConstraints(table, filterDomain)),
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java]: What if there's an overlap between the table properties and the filter? (e.g., a filter limits a column to some range and table properties limit the same column to another range)? If I read this correctly, you'll end up with two entries for that column. This code should do an intersection of the corresponding domains.

**Comment by JamesRTaylor** [comment]
This is ready for review. Thanks for the previous reviews, @martint.

File: presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java [hunk 9]
```diff
@@ -201,17 +203,17 @@ private EstimatedStatsAndCost getEstimatedStatsAndCost()
...
             private final CatalogSchemaTableName table;
-            private final Set<ColumnConstraint> columnConstraints;
+            private final Set<ColumnConstraint> columns;
             private final EstimatedStatsAndCost estimate;
 
             @JsonCreator
             public TableColumnInfo(
                     @JsonProperty("table") CatalogSchemaTableName table,
-                    @JsonProperty("columnConstraints") Set<ColumnConstraint> columnConstraints,
+                    @JsonProperty("columns") Set<ColumnConstraint> columns,
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java]: Why is `columnConstraints` being renamed to `columns`? In particular, changing this in the JSON representation will break backward compatibility.

Comment (JamesRTaylor) [reply presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java]: I renamed it based on feedback you have me here: https://github.com/prestosql/presto/pull/2743#issuecomment-592796177. Maybe I misunderstood? I'm happy to change it back if you think that's better.

Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java]: Ah, sorry if that was confusing... I may have crossed the names. My overall point was that we shouldn't rename it from what it was (and also not add a new field for the filters, but reuse the existing one).

Comment (JamesRTaylor) [reply presto-main/src/main/java/io/prestosql/sql/planner/planprinter/IoPlanPrinter.java]: I've changed the element name back to `columnConstraints`
